### ISSUE TITLE
How to do a Axum's sse function for Candle?

### ISSUE THREAD

**Issue by sunnyregion** [huggingface/candle#2167]
fn run(&mut self, prompt: &str, sample_len: usize) -> Result<()> {
        use std::io::Write;
        self.tokenizer.clear();
        let mut tokens = self
            .tokenizer
            .tokenizer()
            .encode(prompt, true)
            .map_err(E::msg)?
            .get_ids()
            .to_vec();
        for &t in tokens.iter() {
            if let Some(t) = self.tokenizer.next_token(t)? {
                print!("{t}")
            }
        }
        std::io::stdout().flush()?;

        let mut generated_tokens = 0usize;
        let eos_token = match self.tokenizer.get_token("<|endoftext|>") {
            Some(token) => token,
            None => anyhow::bail!("cannot find the <|endoftext|> token"),
        };
        let start_gen = std::time::Instant::now();
        for index in 0..sample_len {
            let context_size = if index > 0 { 1 } else { tokens.len() };
            let start_pos = tokens.len().saturating_sub(context_size);
            let ctxt = &tokens[start_pos..];
            let input = Tensor::new(ctxt, &self.device)?.unsqueeze(0)?;
            let logits = self.model.forward(&input, start_pos)?;
            let logits = logits.squeeze(0)?.squeeze(0)?.to_dtype(DType::F32)?;
            let logits = if self.repeat_penalty == 1. {
                logits
            } else {
                let start_at = tokens.len().saturating_sub(self.repeat_last_n);
                candle_transformers::utils::apply_repeat_penalty(
                    &logits,
                    self.repeat_penalty,
                    &tokens[start_at..],
                )?
            };

            let next_token = self.logits_processor.sample(&logits)?;
            tokens.push(next_token);
            generated_tokens += 1;
            if next_token == eos_token {
                break;
            }
            if let Some(t) = self.tokenizer.next_token(next_token)? {
                print!("{t}");
                std::io::stdout().flush()?;
            }
        }
        let dt = start_gen.elapsed();
        if let Some(rest) = self.tokenizer.decode_rest().map_err(E::msg)? {
            print!("{rest}");
        }
        std::io::stdout().flush()?;
        println!(
            "\n{generated_tokens} tokens generated ({:.2} token/s)",
            generated_tokens as f64 / dt.as_secs_f64(),
        );
        Ok(())
    }

How to rewrite above function to sse?

**Comment by EricLBuehler** [comment]
@sunnyregion, the way we got this to work is by returning a [Sse](https://docs.rs/axum/latest/axum/response/sse/index.html) object containing a streamer (`impl futures:Stream`). At a high level, we use a state object in the streamer to pass messages between token generation and returning the streaming response.

Here is an example of the streamer object:

https://github.com/EricLBuehler/mistral.rs/blob/44c284709642e492df10235fbce1f635a503ce55/mistralrs-server/src/chat_completion.rs#L37-L80

And returning the responder:

https://github.com/EricLBuehler/mistral.rs/blob/44c284709642e492df10235fbce1f635a503ce55/mistralrs-server/src/chat_completion.rs#L236-L246

**Comment by sunnyregion** [comment]
> @sunnyregion, the way we got this to work is by returning a [Sse](https://docs.rs/axum/latest/axum/response/sse/index.html) object containing a streamer (`impl futures:Stream`). At a high level, we use a state object in the streamer to pass messages between token generation and returning the streaming response.
> 
> Here is an example of the streamer object:
> 
> https://github.com/EricLBuehler/mistral.rs/blob/44c284709642e492df10235fbce1f635a503ce55/mistralrs-server/src/chat_completion.rs#L37-L80
> 
> And returning the responder:
> 
> https://github.com/EricLBuehler/mistral.rs/blob/44c284709642e492df10235fbce1f635a503ce55/mistralrs-server/src/chat_completion.rs#L236-L246

thx a lot .I solved the problem.
`pub async fn completion(
    // `Json<T>` will automatically deserialize the request body to a type `T` as JSON.
    Json(body): Json<CompletionRequest>,
) -> Sse<impl Stream<Item = Result<SseEvent, Infallible>>> {
    // `stream!` is a macro from [`async_stream`](https://docs.rs/async-stream/0.3.5/async_stream/index.html) 
    // that makes it easy to create a `futures::stream::Stream` from a generator.
    Sse::new(stream! {
        yield Ok(
          // Create a new `SseEvent` with the default settings.
          // `SseEvent::default().data("Hello, World!")` will return `data: Hello, World!` as the event text chuck.
          SseEvent::default().data(
            …………
            "chat".to_string()
            )
        );
    })
    .keep_alive(KeepAlive::default())
}`
